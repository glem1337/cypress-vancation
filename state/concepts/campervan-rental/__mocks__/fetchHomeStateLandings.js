export default {
 data: {
   data: [
     {
       id: 'cfe698ba-8345-49ba-80b7-aef3a522276a',
       type: 'home_state_landing',
       attributes: {
         state: 'Kansas',
         slug: 'kansas',
       },
     },
   ],
 },
};
