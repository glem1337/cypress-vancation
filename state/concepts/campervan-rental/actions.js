import { CAMPER_INCLUSION } from 'constants/camper';
import { LANDING_TYPE } from 'constants/campervanRentals';
import { SEARCH_RESULTS_RADIUS } from 'constants/searchDestinations';

import {
  FETCH_STATE,
  FETCH_STATE_LOCATION,
  MAP_LOCATION_SLUG_TO_ID,
  MAP_STATE_SLUG_TO_ID,
  FETCH_NEARBY_DESTINATIONS,
  SET_NEARBY_DESTINATIONS_IDS,
  FETCH_EPICENTER_LOCATIONS,
  FETCH_HOME_STATES,
  SET_SHOW_ALL_STATES,
  SET_SLIDE_FAVORITE_DESTINATION,
  SET_FAVORITE_TOTAL,
  FETCH_FAVORITE_DESTINATION,
  FETCH_CAMPERS,
  SET_CAMPER_IDS,
  SET_CAMPERS_PAGE,
  SET_CAMPERS_TOTAL,
  RESET_CAMPERS_DATA,
} from './types';

export const fetchState = ({
  state,
  inclusions = 'seo_info,top_location_landings,fun_facts',
}) => ({
  type: FETCH_STATE,
  state,
  inclusions,
});

export const fetchStateLocation = ({
  state,
  location,
  inclusions = 'state_landing,seo_info,fun_facts',
}) => ({
  type: FETCH_STATE_LOCATION,
  state,
  location,
  inclusions,
});

export const mapStateSlugToId = ({ id, slug }) => ({
  type: MAP_STATE_SLUG_TO_ID,
  id,
  slug,
});

export const mapLocationSlugToId = ({ id, slug }) => ({
  type: MAP_LOCATION_SLUG_TO_ID,
  id,
  slug,
});

export const fetchNearbyDestinations = ({
  latitude,
  longitude,
  count = 12,
  landingType = LANDING_TYPE.LOCATION_LANDING,
  excludedId,
}) => ({
  type: FETCH_NEARBY_DESTINATIONS,
  latitude,
  longitude,
  count,
  landingType,
  excludedId,
});

export const setNearbyDestinationsIds = (ids) => ({
  type: SET_NEARBY_DESTINATIONS_IDS,
  ids,
});

export const fetchEpicenterLocationsAction = () => ({
  type: FETCH_EPICENTER_LOCATIONS,
});

export const fetchHomeStatesAction = () => ({
  type: FETCH_HOME_STATES,
});

export const setShowAllStatesAction = (state) => ({
  type: SET_SHOW_ALL_STATES,
  state,
});

export const setSlideFavoriteDestinationsAction = (slide) => ({
  type: SET_SLIDE_FAVORITE_DESTINATION,
  slide,
});

export const setFavoriteTotalAction = (total) => ({
  type: SET_FAVORITE_TOTAL,
  total,
});

export const fetchFavoriteDestinationsAction = ({ pageNumber, pageSize }) => ({
  type: FETCH_FAVORITE_DESTINATION,
  pageNumber,
  pageSize,
});

export const fetchCampers = ({
  latitude,
  longitude,
  startDate,
  endDate,
  page = 1,
  perPage = 20,
  radius = SEARCH_RESULTS_RADIUS,
  inclusions = CAMPER_INCLUSION.SPECIFICATIONS_DETAILS,
} = {}) => ({
  type: FETCH_CAMPERS,
  latitude,
  longitude,
  startDate,
  endDate,
  page,
  perPage,
  radius,
  inclusions,
});

export const setCamperIds = (ids) => ({
  type: SET_CAMPER_IDS,
  ids,
});

export const setCampersTotal = (total) => ({
  type: SET_CAMPERS_TOTAL,
  total,
});

export const setCampersPage = (page) => ({
  type: SET_CAMPERS_PAGE,
  page,
});

export const resetCampersData = () => ({
  type: RESET_CAMPERS_DATA,
});
