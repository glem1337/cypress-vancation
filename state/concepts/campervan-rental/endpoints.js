import endpoint from 'utils/endpoint';
import {
  locationLandingRoute,
  stateLandingRoute,
  searchNearestLandingPagesRoute,
  epicenterLocationLandingsRoute,
  homeStateLandingsRoute,
  featuredLocationLandingsRoute,
  campersRoute,
} from 'lib/apiRoutes';

import {
  FETCH_STATE,
  FETCH_STATE_LOCATION,
  FETCH_NEARBY_DESTINATIONS,
  FETCH_EPICENTER_LOCATIONS,
  FETCH_HOME_STATES,
  FETCH_FAVORITE_DESTINATION,
  FETCH_CAMPERS,
} from './types';

export const fetchStateEndpoint = endpoint(FETCH_STATE, 'GET', stateLandingRoute);
export const fetchStateLocationEndpoint = endpoint(FETCH_STATE_LOCATION, 'GET', locationLandingRoute);
export const fetchNearbyDestinationsEndpoint = endpoint(FETCH_NEARBY_DESTINATIONS, 'GET', searchNearestLandingPagesRoute);
export const fetchEpicenterLocationLandingsEndpoint = endpoint(FETCH_EPICENTER_LOCATIONS, 'GET', epicenterLocationLandingsRoute);
export const fetchHomeStateLandingsEndpoint = endpoint(FETCH_HOME_STATES, 'GET', homeStateLandingsRoute);
export const favoriteDestinationsEndpoint = endpoint(FETCH_FAVORITE_DESTINATION, 'GET', featuredLocationLandingsRoute);
export const fetchCampersEndpoint = endpoint(FETCH_CAMPERS, 'GET', campersRoute);
