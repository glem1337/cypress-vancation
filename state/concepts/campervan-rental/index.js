import camperVanRentalReducer from 'state/concepts/campervan-rental/reducer';

export { default as camperRentalOperations } from './operations';
export * as camperRentalTypes from './types';

export default camperVanRentalReducer;
