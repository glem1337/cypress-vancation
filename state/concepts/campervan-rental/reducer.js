import { combineReducers } from 'redux';
import { assoc, pathOr } from 'ramda';
import { HYDRATE } from 'next-redux-wrapper';

import { USER_SIGNOUT } from 'state/concepts/session/types';

import {
  MAP_STATE_SLUG_TO_ID,
  MAP_LOCATION_SLUG_TO_ID,
  SET_NEARBY_DESTINATIONS_IDS,
  SET_SHOW_ALL_STATES,
  SET_SLIDE_FAVORITE_DESTINATION,
  SET_FAVORITE_TOTAL,
  SET_CAMPER_IDS,
  SET_CAMPERS_PAGE,
  SET_CAMPERS_TOTAL,
  RESET_CAMPERS_DATA,
} from './types';

const stateIds = (state = {}, action) => {
  switch (action.type) {
    case HYDRATE: {
      return pathOr({}, ['payload', 'campervanRental', 'stateIds'], action);
    }
    case MAP_STATE_SLUG_TO_ID:
      return assoc(action.slug, action.id, state);
    default:
      return state;
  }
};

const locationsIds = (state = {}, action) => {
  switch (action.type) {
    case HYDRATE: {
      return pathOr({}, ['payload', 'campervanRental', 'locationsIds'], action);
    }
    case MAP_LOCATION_SLUG_TO_ID:
      return assoc(action.slug, action.id, state);
    default:
      return state;
  }
};

const nearbyDestinationIds = (state = [], action) => {
  switch (action.type) {
    case SET_NEARBY_DESTINATIONS_IDS:
      return action.ids;
    default:
      return state;
  }
};

const showAllState = (state = false, action) => {
  switch (action.type) {
    case SET_SHOW_ALL_STATES:
      return action.state;
    case HYDRATE:
      return false;
    default:
      return state;
  }
};

const favoriteCurrentSlide = (state = 1, action) => {
  switch (action.type) {
    case SET_SLIDE_FAVORITE_DESTINATION:
      return action.slide;
    case HYDRATE:
      return 1;
    default:
      return state;
  }
};

const favoriteTotal = (state = 1, action) => {
  switch (action.type) {
    case SET_FAVORITE_TOTAL:
      return action.total;
    case HYDRATE:
      return 1;
    default:
      return state;
  }
};

const camperIds = (state = [], action) => {
  switch (action.type) {
    case SET_CAMPER_IDS:
      return action.ids;
    case HYDRATE:
      return action.payload?.campervanRental?.camperIds || [];
    case USER_SIGNOUT: {
      return [];
    }
    case RESET_CAMPERS_DATA: {
      return [];
    }
    default:
      return state;
  }
};

const campersTotal = (state = 0, action) => {
  switch (action.type) {
    case SET_CAMPERS_TOTAL:
      return action.total || 0;
    case HYDRATE:
      return action.payload?.campervanRental?.campersTotal || 0;
    case RESET_CAMPERS_DATA: {
      return 0;
    }
    default:
      return state;
  }
};

const campersPage = (state = 0, action) => {
  switch (action.type) {
    case SET_CAMPERS_PAGE:
      return action.page || 1;
    case HYDRATE:
      return action.payload?.campervanRental?.campersPage || 1;
    case RESET_CAMPERS_DATA: {
      return 0;
    }
    default:
      return state;
  }
};

export default combineReducers({
  stateIds,
  locationsIds,
  nearbyDestinationIds,
  showAllState,
  favoriteCurrentSlide,
  favoriteTotal,
  camperIds,
  campersTotal,
  campersPage,
});
