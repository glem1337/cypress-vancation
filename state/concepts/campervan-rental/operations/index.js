import fetchState from './fetchState';
import fetchStateLocation from './fetchStateLocation';
import fetchNearbyDestinations from './fetchNearbyDestinations';
import fetchHomeStateLandings from './fetchHomeStateLandings';
import fetchEpicenterLocationLandings from './fetchEpicenterLocationLandings';
import fetchFavoriteDestinations from './fetchFavoriteDestinations';
import fetchCampers from './fetchCampers';

export default [
  fetchState,
  fetchStateLocation,
  fetchNearbyDestinations,
  fetchHomeStateLandings,
  fetchEpicenterLocationLandings,
  fetchFavoriteDestinations,
  fetchCampers,
];
