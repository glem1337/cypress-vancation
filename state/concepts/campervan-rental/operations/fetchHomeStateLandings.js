import { createLogic } from 'redux-logic';
import normalize from 'json-api-normalizer';

import { dataApiRequest, dataApiSuccess, dataApiFailure } from 'state/data/actions';
import { FETCH_HOME_STATES } from 'state/concepts/campervan-rental/types';
import { fetchHomeStateLandingsEndpoint } from 'state/concepts/campervan-rental/endpoints';
import showErrorNotifications from 'utils/showErrorNotifications';

const fetchHomeStateLandings = createLogic({
  type: FETCH_HOME_STATES,
  latest: true,

  async process({ httpClient }, dispatch, done) {
    const { endpoint, url } = fetchHomeStateLandingsEndpoint;

    dispatch(dataApiRequest({ endpoint }));

    try {
      const { data } = await httpClient.get(url);

      const response = normalize(data);

      dispatch(dataApiSuccess({ response, endpoint }));
    } catch (error) {
      dispatch(dataApiFailure({ endpoint, error }));
      showErrorNotifications(error, dispatch);
    }
    done();
  },
});

export default fetchHomeStateLandings;
