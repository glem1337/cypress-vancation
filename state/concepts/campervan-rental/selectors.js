import { createSelector } from 'reselect';
import build from 'redux-object';
import * as R from 'ramda';

import isPresent from 'utils/isPresent';
import { loadingSelector } from 'state/data/selectors';

import { fetchCampersEndpoint } from './endpoints';

const dataSelector = R.prop('data');
const stateIdsSelector = R.path(['campervanRental', 'stateIds']);
const locationIdsSelector = R.path(['campervanRental', 'locationsIds']);
const nearbyDestinationIdsSelector = R.path(['campervanRental', 'nearbyDestinationIds']);
const camperIdsSelector = R.path(['campervanRental', 'camperIds']);
export const campersPageSelector = R.path(['campervanRental', 'campersPage']);
export const campersTotalSelector = R.path(['campervanRental', 'campersTotal']);

export const stateLandingSelector = createSelector(
  (_, slug) => slug,
  dataSelector,
  stateIdsSelector,
  (slug, data, ids) => {
    const id = ids[slug];

    return id ? build(data, 'stateLanding', id) : null;
  },
);

export const bindStateLandingSelector = (slug) => (store) => stateLandingSelector(store, slug);

export const locationLandingSelector = createSelector(
  (_, slug) => slug,
  dataSelector,
  locationIdsSelector,
  (slug, data, ids) => {
    const id = ids[slug];

    return id ? build(data, 'locationLanding', id) : null;
  },
);

export const bindLocationLandingSelector = (slug) => (store) => locationLandingSelector(
  store,
  slug,
);

export const nearbyDestinationsSelector = createSelector(
  dataSelector,
  nearbyDestinationIdsSelector,
  (data, ids) => {
    if (R.isEmpty(ids)) {
      return [];
    }

    return R.compose(
      R.filter(item => isPresent(item)),
      R.defaultTo([]),
    )(build(data, 'searchLandingPage', ids));
  },
);

export const campersSelector = createSelector(
  dataSelector,
  camperIdsSelector,
  (data, ids) => {
    if (R.isEmpty(ids)) {
      return [];
    }

    return R.compose(
      R.filter(item => isPresent(item)),
      R.defaultTo([]),
    )(build(data, 'camper', ids));
  },
);

export const areSearchResultsExistedSelector = createSelector(
  camperIdsSelector,
  (ids) => isPresent(ids),
);

export const areSearchResultsFetchingSelector = (state) => loadingSelector(
  state,
  fetchCampersEndpoint.endpoint,
);

export const epicenterLocationsSelector = createSelector(
  dataSelector,
  (data) => build(data, 'epicenterLocationLanding'),
);

export const homeStatesSelector = createSelector(
  dataSelector,
  (data) => build(data, 'homeStateLanding'),
);

export const nearestDestinationsSelector = createSelector(
  dataSelector,
  (data) => build(data, 'searchLandingPage'),
);

export const showAllStateSelector = R.path(['campervanRental', 'showAllState']);

export const locationLandingsSelector = createSelector(
  dataSelector,
  (data) => (build(data, 'locationLanding')),
);

export const favoriteCurrentSlideSelector = R.path(['campervanRental', 'favoriteCurrentSlide']);

export const favoriteTotalSelector = R.path(['campervanRental', 'favoriteTotal']);
