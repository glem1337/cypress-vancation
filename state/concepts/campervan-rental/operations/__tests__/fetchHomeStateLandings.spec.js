import normalize from 'json-api-normalizer';

import { fetchHomeStateLandingsEndpoint } from 'state/concepts/campervan-rental/endpoints';
import { dataApiRequest, dataApiSuccess, dataApiFailure } from 'state/data/actions';
import mockHttpClient from 'utils/testHelpers/mockHttpClient';
import response from 'state/concepts/campervan-rental/__mocks__/fetchHomeStateLandings';
import showErrorNotifications from 'utils/showErrorNotifications';

import fetchHomeStateLandings from '../fetchHomeStateLandings';

jest.mock('utils/showErrorNotifications', () => jest.fn());

describe('fetchHomeStateLandings', () => {
  let dispatch;

  const { endpoint, url } = fetchHomeStateLandingsEndpoint;

  const beforeFunction = httpClient => () => {
    dispatch = jest.fn();
    fetchHomeStateLandings.process({ httpClient }, dispatch, jest.fn());
  };

  it('has valid attributes', () => {
    expect(fetchHomeStateLandings).toMatchSnapshot();
  });

  describe('success', () => {
    const httpClient = mockHttpClient({ method: 'get', response });

    beforeEach(beforeFunction(httpClient));

    it('calls right endpoint', () => {
      expect(httpClient.get).toHaveBeenCalledWith(url);
    });

    it('success', () => {
      expect(dispatch).toHaveBeenCalledTimes(2);

      expect(dispatch).toHaveBeenNthCalledWith(
        1,
        dataApiRequest({ endpoint }),
      );

      expect(dispatch).toHaveBeenNthCalledWith(
        2,
        dataApiSuccess({
          endpoint,
          response: normalize(response.data),
        }),
      );
    });
  });

  describe('failure', () => {
    const error = new Error('test');

    const httpClient = mockHttpClient({ method: 'get', reject: true, response: error });

    beforeEach(beforeFunction(httpClient));

    it('dispatches actions', async () => {
      expect(dispatch).toHaveBeenCalledTimes(2);

      expect(dispatch).toHaveBeenNthCalledWith(
        1,
        dataApiRequest({ endpoint }),
      );

      expect(dispatch).toHaveBeenNthCalledWith(
        2,
        dataApiFailure({ endpoint, error }),
      );
    });

    it('shows error', () => {
      expect(showErrorNotifications).toHaveBeenCalledWith(error, dispatch);
    });
  });
});
