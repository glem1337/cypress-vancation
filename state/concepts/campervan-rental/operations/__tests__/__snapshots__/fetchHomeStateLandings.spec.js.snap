// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchHomeStateLandings has valid attributes 1`] = `
Object {
  "cancelType": undefined,
  "debounce": 0,
  "latest": true,
  "name": undefined,
  "process": [Function],
  "processOptions": Object {
    "dispatchMultiple": true,
  },
  "throttle": 0,
  "transform": undefined,
  "type": "campervan-rentals/FETCH_HOME_STATES",
  "validate": undefined,
  "warnTimeout": 60000,
}
`;
